### PR TITLE
Enhancement: extend hdhomerun widget

### PR DESCRIPTION
**Pull request by JeffRandall**
## Proposed change

 - maintain the previous behavior if no new configuration is done
 - define additional available fields and present the data in order configured
   - global: "tunerCount"
   - per-tuner: "channelNumber", "channelNetwork", "signalStrength", "signalQuality", "symbolQuality", "networkRate", "clientIP"
 - add handling for refreshInterval
 - update docs

An example of two instances of the widget with the fields configured to be `fields: [ "channels", "hd", "tunerCount" ]` and `fields: [ "clientIP", "channelNumber", "channelNetwork", "signalQuality" ]` respectively:

image

Closes #772

## Type of change

- [ ] New service widget
- [ ] Bug fix (non-breaking change which fixes an issue)
- [X] New feature (non-breaking change which adds functionality)
- [ ] Documentation only
- [ ] Other (please explain)

## Checklist:

- [X] If applicable, I have added corresponding documentation changes.
- [X] If applicable, I have reviewed the [feature](https://gethomepage.dev/latest/more/development/#new-feature-guidelines) and / or [service widget guidelines](https://gethomepage.dev/latest/more/development/#service-widget-guidelines).
- [X] I have checked that all code style checks pass using [pre-commit hooks](https://gethomepage.dev/latest/more/development/#code-formatting-with-pre-commit-hooks) and [linting checks](https://gethomepage.dev/latest/more/development/#code-linting).
- [ ] If applicable, I have tested my code for new features & regressions on both mobile & desktop devices, using the latest version of major browsers.

### CODE REVIEW

File: src/widgets/hdhomerun/component.jsx [hunk 1]
```diff
@@ -2,16 +2,68 @@ import Container from "components/services/widget/container";
 import Block from "components/services/widget/block";
 import useWidgetAPI from "utils/proxy/use-widget-api";
 
+function generateDefinitions(channelsData, statusData, tuner) {
```
Comment (shamoon) [inline src/widgets/hdhomerun/component.jsx:5]: While this is an interesting way of doing this it's incompatible with the translation I think. Please rewrite it as other widgets do as a series of <Block> elements etc.

File: src/widgets/hdhomerun/component.jsx [hunk 2]
```diff
@@ -2,16 +2,68 @@ import Container from "components/services/widget/container";
...
+    networkRate: {
+      label: "hdhomerun.networkRate",
+      value: statusData[tuner]?.NetworkRate ?? null,
+    },
+  };
+}
+
 export default function Component({ service }) {
-  const { widget } = service;
+  const { widget, tuner = 0 } = service;
```
Comment (shamoon) [inline src/widgets/hdhomerun/component.jsx:51]: Again, youre retrieving the `tuner` property incorrectly, it should be done like  https://github.com/gethomepage/homepage/blob/a6e4786fc448ce51fe91988a671c3096e3092eca/src/utils/config/service-helpers.js#L352

File: src/widgets/hdhomerun/component.jsx [hunk 3]
```diff
@@ -20,12 +72,17 @@ export default function Component({ service }) {
...
   }
 
-  const hdChannels = channelsData?.filter((channel) => channel.HD === 1);
+  const definitions = generateDefinitions(channelsData, statusData, tuner);
 
   return (
     <Container service={service}>
-      <Block label="hdhomerun.channels" value={channelsData.length} />
-      <Block label="hdhomerun.hd" value={hdChannels.length} />
+      {fields.slice(0, 4).map((field) => (
```
Comment (shamoon) [inline src/widgets/hdhomerun/component.jsx:79]: This fails if you dont specify any fields

File: public/locales/en/common.json [hunk 4]
```diff
@@ -535,7 +535,15 @@
...
-        "hd": "HD"
+        "hd": "HD",
+	"tunerCount": "Tuners",
+	"channelNumber": "Channel",
+	"channelNetwork": "Network",
+	"signalStrength": "Strength",
+	"signalQuality": "Quality",
+	"symbolQuality": "Quality",
+	"networkRate": "Bitrate",
+	"clientIP": "Client"
```
Comment (shamoon) [inline public/locales/en/common.json:546]: Fix indentation please

File: docs/widgets/services/hdhomerun.md [hunk 5]
```diff
@@ -3,12 +3,18 @@ title: HDHomerun
...
+
+If more than 4 fields are provided, only the first 4 are displayed. The
+order that fields are configured is preserved for display.
 
 ```yaml
+tuner: 0 # optional - defaults to 0, used for tuner-specific fields
 widget:
   type: hdhomerun
   url: http://hdhomerun.host.or.ip
+  refreshInterval: 10000 # optional - minimum of 1 sec, default of 10s
```
Comment (shamoon) [inline docs/widgets/services/hdhomerun.md:18]: Is this really necessary? we generally frown upon using the refresh interval. Widgets already update when the user focuses homepage and I dont use HDHomerun but it doesnt look like something that adds a lot of value to be constantly refreshing? Should probably be removed

**Review by shamoon — CHANGES_REQUESTED** [review]
Thanks, this needs a bit of work.

Please also provide sample data for status.json

File: src/widgets/hdhomerun/component.jsx (hunk 2 above)
Comment (JeffRandall) [reply src/widgets/hdhomerun/component.jsx]: Thank you for that pointer.  I was not happy with it being at the top level but didn't see the magic that was required to make it available under the widget.  Fixed.

File: docs/widgets/services/hdhomerun.md [hunk 6]
```diff
@@ -3,12 +3,18 @@ title: HDHomerun
...
+[HDHomerun](https://www.silicondust.com/support/downloads/)
 
-Allowed fields: `["channels", "hd"]`.
+Allowed fields: `["channels", "hd", "tunerCount", "channelNumber", "channelNetwork", "signalStrength", "signalQuality", "symbolQuality, "networkRate", "clientIP" ]`.
+
+If more than 4 fields are provided, only the first 4 are displayed. The
+order that fields are configured is preserved for display.
 
 ```yaml
+tuner: 0 # optional - defaults to 0, used for tuner-specific fields
```
[the remark this answers is not in the document]
Comment (JeffRandall) [reply docs/widgets/services/hdhomerun.md]: Fixed.

File: docs/widgets/services/hdhomerun.md [hunk 7]
```diff
@@ -3,12 +3,18 @@ title: HDHomerun
 description: HDHomerun Widget Configuration
 ---
 
-Learn more about [HDHomerun](https://www.silicondust.com/support/downloads/).
+[HDHomerun](https://www.silicondust.com/support/downloads/)
```
[the remark this answers is not in the document]
Comment (JeffRandall) [reply docs/widgets/services/hdhomerun.md]: Reverted.

File: src/widgets/hdhomerun/component.jsx (hunk 3 above)
Comment (JeffRandall) [reply src/widgets/hdhomerun/component.jsx:79]: Because the fields are provided a default of `["channels", "hd"]` when retrieved (see line 52), fields are always present at this point in the code.

File: public/locales/en/common.json (hunk 4 above)
Comment (JeffRandall) [reply public/locales/en/common.json]: Fixed.

File: docs/widgets/services/hdhomerun.md (hunk 5 above)
Comment (JeffRandall) [reply docs/widgets/services/hdhomerun.md]: Having the page open and it update when someone starts or stops using a tuner without having to click on it was a value-add to me.  This was the reason I set the default fairly high.  I agree that it is not of much value if not using the new fields.

**Comment by JeffRandall** [comment]
Sample status.json from a two tuner hdhomerun:

`[
  {
    "Resource": "tuner0",
    "VctNumber": "15.3",
    "VctName": "TBD",
    "Frequency": 581000000,
    "SignalStrengthPercent": 100,
    "SignalQualityPercent": 77,
    "SymbolQualityPercent": 100,
    "TargetIP": "10.11.2.10",
    "NetworkRate": 1764320
  },
  {
    "Resource": "tuner1"
  }
]`

File: src/widgets/hdhomerun/component.jsx (hunk 1 above)
Comment (JeffRandall) [reply src/widgets/hdhomerun/component.jsx:5]: If I update the entries in `public/locales/en/common.json` and restart, I see the translations change.  For example, 

image
image

Comment (JeffRandall) [reply src/widgets/hdhomerun/component.jsx:5]: The advantage of this approach (which I adapted from customapi doing the mapping in the return statement) is that the user can choose the ordering of the fields in addition to which fields.

File: src/widgets/hdhomerun/component.jsx (hunk 3 above)
Comment (shamoon) [reply src/widgets/hdhomerun/component.jsx:79]: This does not work

```yaml
     - HdHomerun:
          description: hdhomerun
          href: http://localhost
          icon: hdhomerun.png
          widget:
               type: hdhomerun
               url: http://hdhomerun.host.or.ip
```

Screenshot 2024-01-26 at 8 57 33 AM

`fields` will return as `null` not use the default assignment, because of how it is constructed at the container level. Instead you should explicitly check the value of fields (and its length I think) and set it

File: docs/widgets/services/hdhomerun.md [hunk 8]
```diff
@@ -5,10 +5,15 @@ description: HDHomerun Widget Configuration
 
 Learn more about [HDHomerun](https://www.silicondust.com/support/downloads/).
 
-Allowed fields: `["channels", "hd"]`.
+Allowed fields: `["channels", "hd", "tunerCount", "channelNumber", "channelNetwork", "signalStrength", "signalQuality", "symbolQuality, "networkRate", "clientIP" ]`.
```
Comment (shamoon) [inline docs/widgets/services/hdhomerun.md:8]:
```suggestion
Allowed fields: `["channels", "hd", "tunerCount", "channelNumber", "channelNetwork", "signalStrength", "signalQuality", "symbolQuality", "networkRate", "clientIP" ]`.
```

File: src/widgets/hdhomerun/component.jsx [hunk 9]
```diff
@@ -20,12 +72,17 @@ export default function Component({ service }) {
...
+        <Block
+          key={field}
+          label={definitions[Object.keys(definitions).filter((id) => id === field)].label}
+          value={definitions[Object.keys(definitions).filter((id) => id === field)].value}
```
Comment (shamoon) [inline src/widgets/hdhomerun/component.jsx:83]:
```suggestion
          label={definitions[field].label}
          value={definitions[field].value}
```

The larger issue is above, but FYI this is the same

File: src/widgets/hdhomerun/component.jsx (hunk 3 above)
Comment (shamoon) [reply src/widgets/hdhomerun/component.jsx:79]: Ok, youre right translations are OK. The other part of this is that even though this is nicer in some ways I'd like to keep the codebase consistent and all other widgets work the other way (customapi isnt a great example because its so bespoke).

If there came a time we were to use something like this it could be made generalized.

So tl;dr please convert this to something like https://github.com/gethomepage/homepage/blob/19c25713c42c25795e117d2f75d92f69c4211f38/src/widgets/gamedig/component.jsx

(again, I know its less DRY)

File: src/widgets/hdhomerun/component.jsx [hunk 10]
```diff
@@ -2,16 +2,68 @@ import Container from "components/services/widget/container";
...
+  };
+}
+
 export default function Component({ service }) {
   const { widget } = service;
+  const { tuner = 0, refreshInterval = 10000, fields = ["channels", "hd"] } = widget;
 
-  const { data: channelsData, error: channelsError } = useWidgetAPI(widget, "lineup");
+  const { data: channelsData, error: channelsError } = useWidgetAPI(widget, "lineup", {
+    refreshInterval: Math.max(1000, refreshInterval),
```
Comment (shamoon) [inline src/widgets/hdhomerun/component.jsx:55]: Should be removed here too

File: src/widgets/hdhomerun/component.jsx [hunk 11]
```diff
@@ -2,16 +2,68 @@ import Container from "components/services/widget/container";
...
+  });
+  const { data: statusData, error: statusError } = useWidgetAPI(widget, "status", {
+    refreshInterval: Math.max(1000, refreshInterval),
```
Comment (shamoon) [inline src/widgets/hdhomerun/component.jsx:58]: Should be removed here too

**Review by shamoon — CHANGES_REQUESTED** [review]
Thanks for the changes, almost there =)

File: docs/widgets/services/hdhomerun.md (hunk 8 above)
Comment (JeffRandall) [reply docs/widgets/services/hdhomerun.md]: Fixed.

File: src/widgets/hdhomerun/component.jsx (hunk 10 above)
Comment (JeffRandall) [reply src/widgets/hdhomerun/component.jsx]: Removed.

File: src/widgets/hdhomerun/component.jsx (hunk 11 above)
Comment (JeffRandall) [reply src/widgets/hdhomerun/component.jsx]: Removed.

File: docs/widgets/services/hdhomerun.md [hunk 12]
```diff
@@ -5,10 +5,15 @@ description: HDHomerun Widget Configuration
 
 Learn more about [HDHomerun](https://www.silicondust.com/support/downloads/).
 
-Allowed fields: `["channels", "hd"]`.
+Allowed fields: `["channels", "hd", "tunerCount", "channelNumber", "channelNetwork", "signalStrength", "signalQuality", "symbolQuality", "networkRate", "clientIP" ]`.
+
+If more than 4 fields are provided, only the first 4 are displayed. The
+order that fields are configured is preserved for display.
```
Comment (shamoon) [inline docs/widgets/services/hdhomerun.md:11]:
```suggestion
If more than 4 fields are provided, only the first 4 are displayed.
```

**Review by shamoon — APPROVED** [review]
Great thanks for making all the changes

Screenshot 2024-01-26 at 10 54 27 AM
